src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -0,0 +1,249 @@
+using Microsoft.Health.Fhir.SqlServer.Configs;
+using Microsoft.IO;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Storage
+{
+    /// <summary>
+    /// A SQL Server-backed <see cref="IFhirDataStore"/>.
+    /// </summary>
+    public class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
+    {
+        private static readonly Encoding ResourceEncoding = new UnicodeEncoding(bigEndian: false, byteOrderMark: false);
+
+        private readonly SqlServerDataStoreConfiguration _configuration;
+        private readonly SqlServerFhirModel _model;
+        private readonly RecyclableMemoryStreamManager _memoryStreamManager;
+        private readonly ILogger<SqlServerFhirDataStore> _logger;
+
+        public SqlServerFhirDataStore(SqlServerDataStoreConfiguration configuration, SqlServerFhirModel model, ILogger<SqlServerFhirDataStore> logger)
+        {
+            EnsureArg.IsNotNull(configuration, nameof(configuration));
+            EnsureArg.IsNotNull(model, nameof(model));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _configuration = configuration;
+            _model = model;
+            _logger = logger;
+            _memoryStreamManager = new RecyclableMemoryStreamManager();
+        }
+
+        public async Task<UpsertOutcome> UpsertAsync(ResourceWrapper resource, WeakETag weakETag, bool allowCreate, bool keepHistory, CancellationToken cancellationToken)
+        {
+            await _model.EnsureInitialized();
+
+            using (var connection = new SqlConnection(_configuration.ConnectionString))
+            {
+                await connection.OpenAsync(cancellationToken);
+
+                using (var command = connection.CreateCommand())
+                {
+                    command.CommandType = CommandType.StoredProcedure;
+                    command.CommandText = "dbo.UpsertResource";
+
+                    command.Parameters.AddWithValue("@resourceTypeId", _model.GetResourceTypeId(resource.ResourceTypeName));
+                    command.Parameters.Add(new SqlParameter("@resourceId", SqlDbType.VarChar, 64) { Value = resource.ResourceId });
+
+                    int etag = 0;
+                    if (weakETag != null && !int.TryParse(weakETag.VersionId, out etag))
+                    {
+                        throw new ResourceConflictException(weakETag);
+                    }
+
+                    command.Parameters.Add("@eTag", SqlDbType.Int).Value = weakETag == null ? null : (int?)etag;
+                    command.Parameters.AddWithValue("@allowCreate", allowCreate);
+                    command.Parameters.AddWithValue("@isDeleted", resource.IsDeleted);
+                    command.Parameters.AddWithValue("@updatedDateTime", resource.LastModified);
+                    command.Parameters.AddWithValue("@keepHistory", keepHistory);
+                    command.Parameters.AddWithValue("@requestMethod", resource.Request.Method);
```

Comment:
Any reason for not specifying SqlDbType and max length here?

Reply:
It's not critical here. It's important for @resourceId because the parameter is used in a where clause and a mismatch in type can result in an index scan with residual predicate (with a conversion) instead of an index seek. 

So we can do it here as well, but it should not make a bit difference.

Reply:
With codegen, we do the right thing. :)

Reply:
With codegen, we do the right thing. :)